### PR TITLE
ui: Network latency page improvements

### LINKED ISSUE #96101: [ui] Improve Network Latency page to display consistent data during network partition
**Is your feature request related to a problem? Please describe.**
Currently, Network latency page relies on existing data to build latencies matrix and
does lots of filtering/aggregations that is not clear and introduces incorrect 
behavior in cases like network partition.

**Describe the solution you'd like**
To make this page more informative even during network partition, it might require following:
- establish new endpoint that will provide all necessary information regarding network latencies (should allow to reduce calculations on front-end)
- Network Latency page should display information about all known nodes in a cluster, so even for unhealthy nodes;

Jira issue: CRDB-23948

### PR DESCRIPTION
**Pull request by koorosh**
- use connectivity endpoint as a source of connection statuses and latency;
- removed popup with disconnected peers, instead it will be shown
on latency matrix;

Resolves: #96101 

Release note (ui change): add "Show dead nodes" option
to show/hide dead/decommissioned nodes from latency matrix.

Release note (ui change): "No connections" popup menu is removed. Now failed
connections are displayed on the latencies matrix.

- [x] test coverage (storybook added for visual validation)

This PR depends on following PRs that should be merged before this one:
- #95429
- #99191

#### example 1. one node cannot establish connection to another one.
Screenshot 2023-06-26 at 12 33 46

#### example 2. Node 5 is stopped and it is considered as `unavailable` before setting as `dead`
Screenshot 2023-06-26 at 10 52 20

#### example 3. Node 3 is dead. No connection from/to node. Show error message.
Screenshot 2023-06-23 at 14 07 25

#### example 4. Decommissioned node.
 
Screenshot 2023-06-26 at 18 11 23

### CODE REVIEW

**Comment by koorosh** [comment]
сс @dongniwang I added video to show design changes.

**Comment by tbg** [comment]
Drive-by comments, the red in the video for the decommissioned node is surprising to me. "red" means bad but decommissioned nodes are basically "not there", so if anything shouldn't they be greyed out or something like that?

Also want to x-post my comment from [elsewhere](https://github.com/cockroachdb/cockroach/issues/95378#issuecomment-1594580878), it would be great to surface the circuit breaker's connection error in the tooltip if the connection is in a failed state.

**Comment by koorosh** [comment]
Thanks Tobias for feedback, I've updated PR to reflect your suggestions

Updated PR with following
- make tooltip more informative, with error info and indication of connection direction (from, to)
- bring back page auto refreshing every 30sec
- fix sorting dropdown component 
- persist `show dead nodes` option in URI search params

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103837)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh)

___
*[`pkg/ui/workspaces/db-console/src/redux/connectivity.ts` line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/103837#-NYO9osUAhAeBZPdKRbP:-NYO9osUAhAeBZPdKRbQ:b-72q6n0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4960a8701ca87800d09ff71c0977772cef6efec/pkg/ui/workspaces/db-console/src/redux/connectivity.ts#L1)):*
> ```XML
> import { AdminUIState } from "src/redux/state";
> ```

you're getting a lint error here because you're missing the copyright header

**Comment by koorosh** [comment]
Quick demo:
https://www.loom.com/share/9b949e78f26f4254988dee154657f726?sid=6e4a5b39-e1b7-48db-ad60-f52fac3b2f05

**Review by maryliag — APPROVED** [review]
a few nits from me, otherwise the ui code :lgtm: 
but I want to make sure you get an approval from @tbg before merging

> Reviewed 9 of 14 files at r6, 12 of 12 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103837#-:-NZX-JjiCLAK9j3g0QUE:b-q8b9he)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @gtr, @j82w, @koorosh, @THardy98, @xinhaoz, and @zachlite)

___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/network/latency/index.tsx` line 296 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/103837#-NZWz4gz6GR-v6I2jgdS:-NZWz4h-80pngFhN-S6d:b-baheeg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf14ca8153ae1e8b2b34c611367a0260284e8a5/pkg/ui/workspaces/db-console/src/views/reports/containers/network/latency/index.tsx#L296)):*
> ```XML
> 
>   let showError = !!errorMessage;
>   // Show errors for all connection statuses except ESTABLISHING status that treated exceptionally.
> ```

nit: `... status that is treated...`
___
*[`pkg/ui/workspaces/db-console/src/views/app/components/chip/styles.styl` line 42 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/103837#-NZWzZvn1TcX-XbbFK_s:-NZWzZvn1TcX-XbbFK_t:b-4l4w79) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf14ca8153ae1e8b2b34c611367a0260284e8a5/pkg/ui/workspaces/db-console/src/views/app/components/chip/styles.styl#L42)):*
> ```Stylus
>   &--white
>     background $colors--neutral-0
>     border-color #c0c6d9
> ```

there is a color variable for this one that you can use: `$colors--neutral-4`

**Review by tbg — COMMENTED** [review]
Wonderful! I left a review on the Loom. The handling of decommissioned nodes looks like it will cause problems on real-world clusters. The rest is absolutely fantastic and I look forward to this hitting the world.

**Comment by koorosh** [comment]
> I left a review on the Loom.

Thank you. I removed "Show dead nodes" option, so now we show only active nodes.

File: pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx
Comment (xinhaoz) [inline pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx:148]: It looks like this function always uses `this.props`. Can we remove the param and just use `this.props` in the body?

File: pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx
Comment (xinhaoz) [inline pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx:159]: We should avoid refreshing unconditionally in `componentDidUpdate` moving forward (even if cacheddatareducer will implcity throttle these requests). Is there a better, more reliable way or place in the component to issue this refresh? Since this is already a widely used pattern with these db-console specific apis we can live it here for now if there's no alternative.

File: pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx
Comment (xinhaoz) [inline pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx:243]: Can this be lifted out into a separate component?

File: pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx
Comment (xinhaoz) [inline pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx:259]: Why define this component here instead of inlining it with the rest of the JSX?

File: pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx
Comment (xinhaoz) [inline pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx:312]: Suggestion:
```
if (locality === "") { return; }
... rest of the body
```

File: pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx
Comment (xinhaoz) [inline pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx:316]: Same here, change to:
```
if (sort.any(...)) { return; }
... rest
```

File: pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx
Comment (xinhaoz) [inline pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx:304]: This function is quite lengthy with a lot of nested array processing, can you leave a comment summarizing what it's doing?

File: pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx
Comment (xinhaoz) [inline pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx:338]: would an `any` work here?

File: pkg/ui/workspaces/db-console/src/views/reports/containers/network/latency/index.tsx
Comment (xinhaoz) [inline pkg/ui/workspaces/db-console/src/views/reports/containers/network/latency/index.tsx:504]: nit: we should prefer native array functions for uniformity where applicable

**Review by koorosh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103837)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @gtr, @j82w, @maryliag, @THardy98, @xinhaoz, and @zachlite)

___
*[`pkg/ui/workspaces/db-console/src/redux/connectivity.ts` line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/103837#-NYO9osUAhAeBZPdKRbP:-NZe9_0R-RKxsJEx9M8v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4960a8701ca87800d09ff71c0977772cef6efec/pkg/ui/workspaces/db-console/src/redux/connectivity.ts#L1)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

you're getting a lint error here because you're missing the copyright header
</blockquote></details>

Done.
___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx` line 148 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/103837#gh-1253529275:-NZeAL102AioiILJP6EI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4c9a030b3f2bf259c20dbef8082a2af60483f87/pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx#L148)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

It looks like this function always uses `this.props`. Can we remove the param and just use `this.props` in the body?
</blockquote></details>

Done.
___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx` line 159 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/103837#gh-1253534020:-NZeAb2R-PgqV-hk7nMW:b1v7mmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4c9a030b3f2bf259c20dbef8082a2af60483f87/pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx#L159)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

We should avoid refreshing unconditionally in `componentDidUpdate` moving forward (even if cacheddatareducer will implcity throttle these requests). Is there a better, more reliable way or place in the component to issue this refresh? Since this is already a widely used pattern with these db-console specific apis we can live it here for now if there's no alternative.
</blockquote></details>

I tend to keep it as is to avoid potential regressions and avoid significant component refactoring in scope of this change.
This component isn't perfect at all and IMO needs lots of refactoring to be well structured and easy to maintain.
I created a separate https://github.com/cockroachdb/cockroach/issues/106244 to address such improvements.
___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx` line 243 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/103837#gh-1253535753:-NZeEmdL7KnkSJ00xYsx:b-r3i9kr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4c9a030b3f2bf259c20dbef8082a2af60483f87/pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx#L243)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Can this be lifted out into a separate component?
</blockquote></details>

Same as comment above. This component shares state so it's not just trivial extraction of pure component. 
___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx` line 259 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/103837#gh-1253537921:-NZeFSiBByg-I5aa3DZk:b-tpi9pe) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4c9a030b3f2bf259c20dbef8082a2af60483f87/pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx#L259)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Why define this component here instead of inlining it with the rest of the JSX?
</blockquote></details>

It is used twice depending on condition, avoids code duplication.
___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx` line 304 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/103837#gh-1253548651:-NZeQQZAC_qM5ekFT9Dm:b-kabw27) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4c9a030b3f2bf259c20dbef8082a2af60483f87/pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx#L304)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

This function is quite lengthy with a lot of nested array processing, can you leave a comment summarizing what it's doing?
</blockquote></details>

Definitely! 👍🏼
___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx` line 312 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/103837#gh-1253541543:-NZeRAml4vDNGzKqurUw:b-ara5r2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4c9a030b3f2bf259c20dbef8082a2af60483f87/pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx#L312)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Suggestion:
```
if (locality === "") { return; }
... rest of the body
```
</blockquote></details>

I like this! Done.
___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx` line 316 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/103837#gh-1253545411:-NZeRewV5oXoEuWhRxRh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4c9a030b3f2bf259c20dbef8082a2af60483f87/pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx#L316)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Same here, change to:
```
if (sort.any(...)) { return; }
... rest
```
</blockquote></details>

Done.
___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx` line 338 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/103837#gh-1253551351:-NZeTh7w49d1ugKVHVBk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4c9a030b3f2bf259c20dbef8082a2af60483f87/pkg/ui/workspaces/db-console/src/views/reports/containers/network/index.tsx#L338)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

would an `any` work here?
</blockquote></details>

Done.
___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/network/latency/index.tsx` line 296 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/103837#-NZWz4gz6GR-v6I2jgdS:-NZeTvp07IYvaehiSgQm:b-i21bbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf14ca8153ae1e8b2b34c611367a0260284e8a5/pkg/ui/workspaces/db-console/src/views/reports/containers/network/latency/index.tsx#L296)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

nit: `... status that is treated...`
</blockquote></details>

Thanks. Done.
___
*[`pkg/ui/workspaces/db-console/src/views/reports/containers/network/latency/index.tsx` line 504 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/103837#gh-1253575710:-NZeTz_qAokuYaARdpd6:b-ckn8fr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4c9a030b3f2bf259c20dbef8082a2af60483f87/pkg/ui/workspaces/db-console/src/views/reports/containers/network/latency/index.tsx#L504)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

nit: we should prefer native array functions for uniformity where applicable
</blockquote></details>

These components heavily use lodash in many places. Prefer to change it in scope of #  #106244
___
*[`pkg/ui/workspaces/db-console/src/views/app/components/chip/styles.styl` line 42 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/103837#-NZWzZvn1TcX-XbbFK_s:-NZeVNXPDKvCCys9WTIq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf14ca8153ae1e8b2b34c611367a0260284e8a5/pkg/ui/workspaces/db-console/src/views/app/components/chip/styles.styl#L42)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

there is a color variable for this one that you can use: `$colors--neutral-4`
</blockquote></details>

Done.

**Review by maryliag — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103837#-:-N__lx8U6Y7c0xO-ciIh:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @gtr, @j82w, @THardy98, @xinhaoz, and @zachlite)

**Pull request merged**
